src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs

```diff
@@ -95,85 +94,51 @@ private BoundExpression BindWithExpression(WithExpressionSyntax syntax, Diagnost
-                        if (lookupResult.IsSingleViable &&
-                            lookupResult.SingleSymbolOrDefault is var sym)
-                        {
-                            switch (sym.Kind)
-                            {
-                                case SymbolKind.Property:
-                                    member = sym;
-                                    // PROTOTYPE: this should check for init-only, but that isn't a separate feature yet
-                                    // It also will not work in metadata.
-                                    if (!(sym is SynthesizedRecordPropertySymbol))
-                                    {
-                                        goto default;
-                                    }
-                                    break;
-
-                                default:
-                                    hasErrors = true;
-                                    diagnostics.Add(
-                                        ErrorCode.ERR_WithMemberIsNotRecordProperty,
-                                        location);
-                                    break;
-                            }
-                        }
-
-                        if (!hasErrors && member is null)
-                        {
-                            hasErrors = true;
-                            Error(
-                                diagnostics,
-                                ErrorCode.ERR_NoSuchMemberOrExtension,
-                                location,
-                                cloneReturnType,
-                                propName);
-                        }
-                    }
-
-                    boundRight = BindValue(assignment.Right, diagnostics, BindValueKind.RValue);
-                    if (!(member is null))
-                    {
-                        boundRight = GenerateConversionForAssignment(
-                            member.GetTypeOrReturnType().Type,
-                            boundRight,
-                            diagnostics);
-                    }
-                    lookupResult.Clear();
+                    BoundExpression? boundMember = null;
+                    boundMember = BindInstanceMemberAccess(
+                        node: left,
+                        right: left,
+                        boundLeft: implicitReceiver,
+                        rightName: propName,
+                        rightArity: 0,
+                        typeArgumentsSyntax: default(SeparatedSyntaxList<TypeSyntax>),
+                        typeArgumentsWithAnnotations: default(ImmutableArray<TypeWithAnnotations>),
+                        invoked: false,
+                        indexed: false,
+                        diagnostics: diagnostics);
+
+                    hasErrors |= boundMember.HasAnyErrors || implicitReceiver.HasAnyErrors;
+
+                    boundMember = CheckValue(boundMember, BindValueKind.Assignable, diagnostics);
```

Comment:
>CheckValue [](start = 34, length = 10)

I don't see how this enforces the init rules. How does CheckValue know that this is one of the contexts in which assignment to an init-only member is permitted?

Reply:
`CheckPropertyValueKind` checks the receiver of the access:
```
           bool isAllowedInitOnlySet(BoundExpression receiver)
            {
                // ok: new C() { InitOnlyProperty = ... }
                if (receiver is BoundObjectOrCollectionValuePlaceholder)
                {
                    return true;
                }
...
```

---
In reply to: [431489149](https://github.com/dotnet/roslyn/pull/44582#discussion_r431489149) [](ancestors = 431489149)